### ISSUE TITLE
[java] UnusedPrivateMethod false positive from inner class via external class

### ISSUE THREAD

**Issue by arifogel** [pmd/pmd#1189]
**Affects PMD Version: 6.4.0** 

**Rule: UnusedPrivateMethod**

**Description:**
A private method of an outer class called in an inner class via a method from an external class returning an object of the outer class type is falsely reported as being unused.

**Code Sample demonstrating the issue:**
- clone https://github.com/arifogel/pmd-unused-private-method-test
- run `mvn install` in root of downloaded repository

**Running PMD through:** *[Maven]*

**Comment by arifogel** [comment]
[output.log](https://github.com/pmd/pmd/files/2100858/output.log)

**Comment by jsotuyod** [comment]
@arifogel thanks for the report. The rule is effectively not using type resolution yet, so these non-straightforward scenarios will be missed for the time being.

**Comment by arifogel** [comment]
I'd be willing to look into this with a little shepherding. Is there some documentation/code I could review to help me get started? Or is this blocked on something that is currently churning internally?

**Comment by arifogel** [comment]
@jsotuyod any progress on this? I'm still willing to look into this with a bit of initial guidance.

**Comment by adangel** [comment]
Hi @arifogel,
first step would be (if not done already): https://github.com/pmd/pmd/wiki/Setup-IDE
Then you can add your case as a new test case into https://github.com/pmd/pmd/blob/master/pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/UnusedPrivateMethod.xml  (you would add here only your `OuterClass`). Since you are dealing with type resolution, you'll need additionally the two classes as java source, so that it is compiled and available during test runtime. You can add your two classes `OuterClass` and `Referrer` to the package `net.sourceforge.pmd.lang.java.rule.bestpractices` under `src/test/java`.
You can start the unit test then through `BestPracticesRulesTest`. You should now see a failure. You can also run the test in the debugger and step through the class https://github.com/pmd/pmd/blob/master/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/UnusedPrivateMethodRule.java .

Then the hard part comes: Fixing the issue. We currently don't resolve the return type of a method call - which would be needed here in order to know, that we call the method `privateMethod()`.

I guess, changing the sample code to this one:

```
public final class OuterClass {
  private static class InnerClass {
    private void callPrivateMethod() {
      OuterClass outer = Referrer.getOuterClass();
      outer.privateMethod();
    }
  }

  private void privateMethod() {}
}
```

won't trigger this false positive, since we know the type of the variable `outer`. If this also triggers the rule, then this might be a first easier step to fix.

Let me know, if you run into any problems - either here or on https://gitter.im/pmd/pmd

**Comment by arifogel** [comment]
I finally got my IDE set up. Your version of the sample code passes.
I was able to condense the false positive test code to a single file:
```
public final class OuterClass {
    public static class InnerClass {
        public void callPrivateMethod() {
            Referrer.getOuterClass().privateMethod();
        }
    }

    public static class Referrer {
        public static OuterClass getOuterClass() {
            return new OuterClass();
        }
    }

    private void privateMethod() {
    }
}
```
Now comes the hard part of processing the return type..

**Comment by arifogel** [comment]
It seems like the failure occurs even earlier than I expected.
I changed `net.sourceforge.pmd.lang.java.symboltable.TRACE` to `true` and ran using pmd master for the concise snippet above.
The trace is unexpected:
```
new search for variable Referrer:4:class net.sourceforge.pmd.lang.java.ast.ASTName
 checking scope MethodScope: for name occurrence Referrer:4:class net.sourceforge.pmd.lang.java.ast.ASTName
 moving up from MethodScope: to ClassScope (InnerClass): Inner Classes ; Method callPrivateMethod, line 3, params = 0(begins at line 3, 0 usages), 
 checking scope ClassScope (InnerClass): Inner Classes ; Method callPrivateMethod, line 3, params = 0(begins at line 3, 0 usages),  for name occurrence Referrer:4:class net.sourceforge.pmd.lang.java.ast.ASTName
 moving up from ClassScope (InnerClass): Inner Classes ; Method callPrivateMethod, line 3, params = 0(begins at line 3, 0 usages),  to ClassScope (OuterClass): Method privateMethod, line 14, params = 0(begins at line 14, 0 usages), 
 checking scope ClassScope (OuterClass): Method privateMethod, line 14, params = 0(begins at line 14, 0 usages),  for name occurrence Referrer:4:class net.sourceforge.pmd.lang.java.ast.ASTName
 moving up from ClassScope (OuterClass): Method privateMethod, line 14, params = 0(begins at line 14, 0 usages),  to SourceFileScope: Class OuterClass
 checking scope SourceFileScope: Class OuterClass for name occurrence Referrer:4:class net.sourceforge.pmd.lang.java.ast.ASTName
found []
```
AFAICT, even `Referrer` by itself cannot be found. Is this the same problem of not using type resolution, or is this a separate issue?

**Comment by KroArtem** [comment]
It seems like we also encountered this issue.

**Comment by adangel** [comment]
This has been fixed with PMD 7.0.0-rc1.
